src/cares_wrap.cc

```diff
@@ -637,7 +649,7 @@ class QueryWrap : public AsyncWrap {
   static void CaresAsyncCb(uv_async_t* handle) {
     auto data = static_cast<struct CaresAsyncData*>(handle->data);
 
-    QueryWrap* wrap = data->wrap;
+    BaseWrap* wrap = data->wrap;
```

Comment:
@SirR4T It’s hard to tell until this PR is at the point where the option starts to work.

I’d personally not split `BaseWrap` into two `QueryWrap` and `SearchWrap`, and instead pass the new constructor parameter along from the subclasses, e.g.:

```c++

class QueryAnyWrap: public QueryWrap {
 public:
  QueryAnyWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
    : QueryWrap(channel, req_wrap_obj) {
  }
```

becomes

```c++

class QueryAnyWrap: public QueryWrap {
 public:
  QueryAnyWrap(ChannelWrap* channel, Local<Object> req_wrap_obj, ares_function* fn)
    : QueryWrap(channel, req_wrap_obj, fn) {
  }
```

and then decide which function to pass in at the call site (the `Query` function). But yeah, it’s hard to tell in the current state of things…

Reply:
@addaleax should 

```
  env->SetProtoMethod(channel_wrap, "queryAny", Query<QueryAnyWrap>);
  env->SetProtoMethod(channel_wrap, "queryA", Query<QueryAWrap>);
  env->SetProtoMethod(channel_wrap, "queryAaaa", Query<QueryAaaaWrap>);
  env->SetProtoMethod(channel_wrap, "queryCname", Query<QueryCnameWrap>);
  env->SetProtoMethod(channel_wrap, "queryMx", Query<QueryMxWrap>);
  env->SetProtoMethod(channel_wrap, "queryNs", Query<QueryNsWrap>);
  env->SetProtoMethod(channel_wrap, "queryTxt", Query<QueryTxtWrap>);
  env->SetProtoMethod(channel_wrap, "querySrv", Query<QuerySrvWrap>);
  env->SetProtoMethod(channel_wrap, "queryPtr", Query<QueryPtrWrap>);
  env->SetProtoMethod(channel_wrap, "queryNaptr", Query<QueryNaptrWrap>);
  env->SetProtoMethod(channel_wrap, "querySoa", Query<QuerySoaWrap>);
  env->SetProtoMethod(channel_wrap, "getHostByAddr", Query<GetHostByAddrWrap>);
```

become something like this:
```
  env->SetProtoMethod(channel_wrap, "queryAny", Query<QueryAnyWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryA", Query<QueryAWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryAaaa", Query<QueryAaaaWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryCname", Query<QueryCnameWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryMx", Query<QueryMxWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryNs", Query<QueryNsWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryTxt", Query<QueryTxtWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "querySrv", Query<QuerySrvWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryPtr", Query<QueryPtrWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "queryNaptr", Query<QueryNaptrWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "querySoa", Query<QuerySoaWrap, ares_query>);
  env->SetProtoMethod(channel_wrap, "getHostByAddr", Query<GetHostByAddrWrap, ares_query>);

  env->SetProtoMethod(channel_wrap, "searchAny", Query<QueryAnyWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchA", Query<QueryAWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchAaaa", Query<QueryAaaaWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchCname", Query<QueryCnameWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchMx", Query<QueryMxWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchNs", Query<QueryNsWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchTxt", Query<QueryTxtWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchSrv", Query<QuerySrvWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchPtr", Query<QueryPtrWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchNaptr", Query<QueryNaptrWrap, ares_search>);
  env->SetProtoMethod(channel_wrap, "searchSoa", Query<QuerySoaWrap, ares_search>);
```
?

Although that kind of seems to defeat the purpose of passing in a `search: true` option to C++ land.

At this point, I think I need some pointers on how to pass the `search: true` option into (and parse for, while in) C++ land.

Reply:
@SirR4T Yes, that’s what I want to avoid… The suggestion from my comment above would allow you to inspect the `search` option inside the `Query<…>` function, and then pass `ares_search` or `ares_query` depending on whether it is set or not

Reply:
yay 🎉 

build passed, as well as js tests passed, and removed `SearchWrap` as well.